packages/flutter_tools/test/general.shard/resident_runner_test.dart

```diff
@@ -53,7 +53,7 @@ void main() {
   setUp(() {
     testbed = Testbed(setup: () {
       fakeAnalytics = getInitializedFakeAnalyticsInstance(
-        fs: globals.fs,
+        fs: MemoryFileSystem.test(),
```

Comment:
Why did you make this change?

Reply:
And it turns out that globals.fs is not always a MemoryFileSystem for this test which is why i swapped it out for a MemoryFileSystem...  this works for the test because we don't touch the filesystem within the test

Reply:
I see. I will have to think about this. This code seems non-ideal because we now have multiple file systems in a single test.

Reply:
Is globals.fs a MemoryFileSystem? If so, maybe we should just do a cast.

Reply:
Yeah i was surprised myself to see that we were referencing globals within the tests for that file, but i tried to do a typecast the first time, the [logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8749844273835272737/+/u/run_test.dart_for_tool_tests_shard_and_subshard_general/stdout) showed that it was sometimes defaulting to the original fs `ErrorHandlingFileSystem`

```
00:19 +804: test/general.shard/resident_runner_test.dart: ResidentRunner reports error with missing entrypoint file
type 'ErrorHandlingFileSystem' is not a subtype of type 'MemoryFileSystem' in type cast
#0      main.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter_tools/test/general.shard/resident_runner_test.dart:56:24)
#1      Testbed.run.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter_tools/test/src/testbed.dart:129:28)
#2      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:29)
#3      _rootRun (dart:async/zone.dart:1399:13)
#4      _CustomZone.run (dart:async/zone.dart:1301:19)
#5      _runZoned (dart:async/zone.dart:1826:10)
#6      runZoned (dart:async/zone.dart:1763:10)
#7      AppContext.run (package:flutter_tools/src/base/context.dart:152:12)
#8      Testbed.run.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter_tools/test/src/testbed.dart:111:24)
#9      runInContext.runnerWrapper (package:flutter_tools/src/context_runner.dart:82:18)
<asynchronous suspension>
#10     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#11     testUsingContext.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter_tools/test/src/context.dart:151:26)
<asynchronous suspension>
#12     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:19)
<asynchronous suspension>
#13     testUsingContext.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter_tools/test/src/context.dart:140:22)
<asynchronous suspension>

00:19 +804 -1: test/general.shard/resident_runner_test.dart: ResidentRunner reports error with missing entrypoint file [E]
  type 'ErrorHandlingFileSystem' is not a subtype of type 'MemoryFileSystem' in type cast
  test/general.shard/resident_runner_test.dart 56:24   main.<fn>.<fn>
  test/src/testbed.dart 129:28                         Testbed.run.<fn>.<fn>.<fn>
  package:flutter_tools/src/base/context.dart 153:29   AppContext.run.<fn>
  dart:async                                           runZoned
  package:flutter_tools/src/base/context.dart 152:12   AppContext.run
  test/src/testbed.dart 111:24                         Testbed.run.<fn>.<fn>
  package:flutter_tools/src/context_runner.dart 82:18  runInContext.runnerWrapper
  ===== asynchronous gap ===========================
  package:flutter_tools/src/base/context.dart 153:19   AppContext.run.<fn>
  ===== asynchronous gap ===========================
  test/src/context.dart 151:26                         testUsingContext.<fn>.<fn>.<fn>.<fn>.<fn>
  ===== asynchronous gap ===========================
  package:flutter_tools/src/base/context.dart 153:19   AppContext.run.<fn>
  ===== asynchronous gap ===========================
  test/src/context.dart 140:22                         testUsingContext.<fn>.<fn>.<fn>.<fn>
```

Reply:
Is the errorhandlingfilesystem wrapping a memoryfilesystem?

Reply:
I think it is possible to pass in a FileSystem override that points to MemoryFileSystem, but then the code gets a little messy but works for `test/general.shard/isolated/resident_runner_test.dart`

```dart
      fakeAnalytics = getInitializedFakeAnalyticsInstance(
        fs: (globals.fs as ErrorHandlingFileSystem).fileSystem as MemoryFileSystem,
        fakeFlutterVersion: FakeFlutterVersion(),
      );
```

We would be overriding the delegate within `ErrorHandlingFileSystem`

Reply:
I'm going to check out this branch and take a closer look

Reply:
I pushed a commit where the two testbed test libraries were re-written to just not use `FakeAnalytics` from package:unified_analytics: https://github.com/flutter/flutter/pull/147250/commits/bc5901323a11629aeb65b4b4447230f475317905

Reply:
More generally (and outside the scope of this PR), it might be advantageous in the tool if we just stopped importing `FakeAnalytics` from package:unified_analytics, if breaking changes to that class is going to necessitate these kinds of refactors.

Even more generally, we should try at all costs to avoid having duplicate instances of globals within a single test run, as that leads to very difficult to reason about correctness bugs. This is especially easy to do with the `TestBed`, which is one of the reasons we stopped using it.

Reply:
>I pushed a commit where the two testbed test libraries were re-written to just not use FakeAnalytics from package:unified_analytics: https://github.com/flutter/flutter/commit/bc5901323a11629aeb65b4b4447230f475317905

SGTM, this makes sense for those files since we aren't actually testing if an event was sent using the built in logic in package:unified_analytics

>Even more generally, we should try at all costs to avoid having duplicate instances of globals within a single test run

I also agree with this, the change I had pushed most recently was using only one FileSystem, it was just a bit more work to get the delegate if we used the `ThrowingForwardingFileSystem` in that one test

> More generally (and outside the scope of this PR), it might be advantageous in the tool if we just stopped importing FakeAnalytics from package:unified_analytics

I think there is value in using the exported `FakeAnalytics` because there may be tests that rely on the logic for package:unified_analytics to be correct.  I do agree though that we don't really need to be using `FakeAnalytics` for tests that don't check for events sent, we could probably swap those with the `NoOpAnalytics` since we usually just add it to class constructor being tested

Reply:
> there may be tests that rely on the logic for package:unified_analytics to be correct

Can you give concrete examples?

> I do agree though that we don't really need to be using FakeAnalytics for tests that don't check for events sent

If we want to know which events were sent, we can use the fake implementation I have here, right? Or is there additional behavior we want to test for that we would not want to re-implement here?

Reply:
Let me just file a separate issue where we can discuss independent of this roll :)